File: examples/src/main/java/com/mapbox/navigation/examples/MainActivity.java
package com.mapbox.navigation.examples;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.mapbox.android.core.permissions.PermissionsListener;
import com.mapbox.android.core.permissions.PermissionsManager;
import com.mapbox.navigation.examples.activity.HybridNavigationActivity;
import com.mapbox.navigation.examples.activity.MockNavigationActivity;
import com.mapbox.navigation.examples.activity.OffboardRouterActivityJava;
import com.mapbox.navigation.examples.activity.OffboardRouterActivityKt;
import com.mapbox.navigation.examples.activity.OnboardRouterActivityJava;
import com.mapbox.navigation.examples.activity.OnboardRouterActivityKt;
import com.mapbox.navigation.examples.activity.SimpleMapboxNavigationKt;
import com.mapbox.navigation.examples.activity.TripServiceActivityKt;
import com.mapbox.navigation.examples.activity.TripSessionActivityKt;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;

public class MainActivity extends AppCompatActivity implements PermissionsListener {

  @BindView(R.id.settingsFab)
  FloatingActionButton settingsFab;

  private static final int CHANGE_SETTING_REQUEST_CODE = 1;
  private RecyclerView recyclerView;
  private PermissionsManager permissionsManager;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    ButterKnife.bind(this);

    settingsFab.setOnClickListener(v -> startActivityForResult(
            new Intent(MainActivity.this, NavigationSettingsActivity.class),
            CHANGE_SETTING_REQUEST_CODE
    ));

    final List<SampleItem> samples = new ArrayList<>(Arrays.asList(
      new SampleItem(
        getString(R.string.title_simple_navigation_kotlin),
        getString(R.string.description_simple_navigation_kotlin),
        SimpleMapboxNavigationKt.class
      ),
      new SampleItem(
        getString(R.string.title_mock_navigation),
        getString(R.string.description_mock_navigation),
        MockNavigationActivity.class
      ),
      new SampleItem(
        getString(R.string.title_offboard_router_kotlin),
        getString(R.string.description_offboard_router_kotlin),
        OffboardRouterActivityKt.class
      ),
      new SampleItem(
        getString(R.string.title_component_hybrid_navigation),
        getString(R.string.description_hybrid_router),
        HybridNavigationActivity.class
      ),
      new SampleItem(
        getString(R.string.title_offboard_router_java),
        getString(R.string.description_offboard_router_java),
        OffboardRouterActivityJava.class
      ),
      new SampleItem(
        getString(R.string.title_onboard_router_kotlin),
        getString(R.string.description_onboard_router_kotlin),
        OnboardRouterActivityKt.class
      ),
      new SampleItem(
        getString(R.string.title_onboard_router_java),
        getString(R.string.description_onboard_router_java),
        OnboardRouterActivityJava.class
      ),
      new SampleItem(
        getString(R.string.title_trip_service_kotlin),
        getString(R.string.description_trip_service_kotlin),
        TripServiceActivityKt.class
      ),
      new SampleItem(
        getString(R.string.title_trip_session_kotlin),
        getString(R.string.description_trip_session_kotlin),
        TripSessionActivityKt.class
      )
    ));

    // RecyclerView
    recyclerView = findViewById(R.id.recycler_view);
    recyclerView.setHasFixedSize(true);

    // Use a linear layout manager
    RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
    recyclerView.setLayoutManager(layoutManager);

    // Specify an adapter
    RecyclerView.Adapter adapter = new MainAdapter(samples);
    recyclerView.setAdapter(adapter);

    // Check for location permission
    permissionsManager = new PermissionsManager(this);
    if (!PermissionsManager.areLocationPermissionsGranted(this)) {
      recyclerView.setVisibility(View.INVISIBLE);
      permissionsManager.requestLocationPermissions(this);
    } else {
      requestPermissionIfNotGranted(WRITE_EXTERNAL_STORAGE);
    }
  }

  @Override
  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                         @NonNull int[] grantResults) {
    if (requestCode == 0) {
      permissionsManager.onRequestPermissionsResult(requestCode, permissions, grantResults);
    } else {
      boolean granted = grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED;
      if (!granted) {
        recyclerView.setVisibility(View.INVISIBLE);
        Toast.makeText(this, "You didn't grant storage permissions.", Toast.LENGTH_LONG).show();
      } else {
        recyclerView.setVisibility(View.VISIBLE);
      }
    }
  }

  @Override
  public void onExplanationNeeded(List<String> permissionsToExplain) {
    Toast.makeText(this, "This app needs location and storage permissions"
      + "in order to show its functionality.", Toast.LENGTH_LONG).show();
  }

  @Override
  public void onPermissionResult(boolean granted) {
    if (granted) {
      requestPermissionIfNotGranted(WRITE_EXTERNAL_STORAGE);
    } else {
      Toast.makeText(this, "You didn't grant location permissions.", Toast.LENGTH_LONG).show();
    }
  }

  private void requestPermissionIfNotGranted(String permission) {
    List<String> permissionsNeeded = new ArrayList<>();
    if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
      permissionsNeeded.add(permission);
      ActivityCompat.requestPermissions(this, permissionsNeeded.toArray(new String[0]), 10);
    }
  }

  /*
   * Recycler view
   */

  private class MainAdapter extends RecyclerView.Adapter<MainAdapter.ViewHolder> {

    private List<SampleItem> samples;

    class ViewHolder extends RecyclerView.ViewHolder {

      private TextView nameView;
      private TextView descriptionView;

      ViewHolder(View view) {
        super(view);
        nameView = view.findViewById(R.id.nameView);
        descriptionView = view.findViewById(R.id.descriptionView);
      }
    }

    MainAdapter(List<SampleItem> samples) {
      this.samples = samples;
    }

    @NonNull
    @Override
    public MainAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
      View view = LayoutInflater
        .from(parent.getContext())
        .inflate(R.layout.item_main_feature, parent, false);

      view.setOnClickListener(clickedView -> {
        int position = recyclerView.getChildLayoutPosition(clickedView);
        Intent intent = new Intent(clickedView.getContext(), samples.get(position).getActivity());
        startActivity(intent);
      });

      return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MainAdapter.ViewHolder holder, int position) {
      holder.nameView.setText(samples.get(position).getName());
      holder.descriptionView.setText(samples.get(position).getDescription());
    }

    @Override
    public int getItemCount() {
      return samples.size();
    }
  }
}

File: examples/src/main/java/com/mapbox/navigation/examples/MainActivity.kt
package com.mapbox.navigation.examples

import android.Manifest.permission
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.mapbox.android.core.permissions.PermissionsListener
import com.mapbox.common.LogConfiguration
import com.mapbox.common.LoggingLevel
import com.mapbox.navigation.examples.core.IndependentRouteGenerationActivity
import com.mapbox.navigation.examples.core.MapboxBuildingHighlightActivity
import com.mapbox.navigation.examples.core.MapboxCustomStyleActivity
import com.mapbox.navigation.examples.core.MapboxJunctionActivity
import com.mapbox.navigation.examples.core.MapboxManeuverActivity
import com.mapbox.navigation.examples.core.MapboxMultipleArrowActivity
import com.mapbox.navigation.examples.core.MapboxNavigationActivity
import com.mapbox.navigation.examples.core.MapboxRouteLineAndArrowActivity
import com.mapbox.navigation.examples.core.MapboxSignboardActivity
import com.mapbox.navigation.examples.core.MapboxTripProgressActivity
import com.mapbox.navigation.examples.core.MapboxVoiceActivity
import com.mapbox.navigation.examples.core.MultiLegRouteExampleActivity
import com.mapbox.navigation.examples.core.R
import com.mapbox.navigation.examples.core.ReplayHistoryActivity
import com.mapbox.navigation.examples.core.camera.MapboxCameraAnimationsActivity
import com.mapbox.navigation.examples.core.databinding.LayoutActivityMainBinding
import com.mapbox.navigation.examples.util.LocationPermissionsHelper
import com.mapbox.navigation.examples.util.LocationPermissionsHelper.Companion.areLocationPermissionsGranted
import com.mapbox.navigation.examples.util.RouteDrawingActivity

class MainActivity : AppCompatActivity(), PermissionsListener {

    private val locationPermissionsHelper = LocationPermissionsHelper(this)
    private lateinit var binding: LayoutActivityMainBinding
    private lateinit var adapter: ExamplesAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        LogConfiguration.setLoggingLevel(LoggingLevel.DEBUG)
        binding = LayoutActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val sampleItemList = buildSampleList()
        adapter = ExamplesAdapter(this) {
            startActivity(Intent(this@MainActivity, sampleItemList[it].activity))
        }
        binding.coreRecycler.apply {
            layoutManager = LinearLayoutManager(
                this@MainActivity,
                LinearLayoutManager.VERTICAL,
                false
            )
            adapter = this@MainActivity.adapter
        }
        adapter.addSampleItems(sampleItemList)

        if (areLocationPermissionsGranted(this)) {
            requestOptionalPermissions()
        } else {
            locationPermissionsHelper.requestLocationPermissions(this)
        }
    }

    private fun buildSampleList(): List<SampleItem> {
        return listOf(
            SampleItem(
                getString(R.string.title_navigation),
                getString(R.string.description_navigation),
                MapboxNavigationActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_route_line),
                getString(R.string.description_route_line),
                MapboxRouteLineAndArrowActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_camera),
                getString(R.string.description_camera),
                MapboxCameraAnimationsActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_signboard),
                getString(R.string.description_signboard),
                MapboxSignboardActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_junction),
                getString(R.string.description_junction),
                MapboxJunctionActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_trip_progress),
                getString(R.string.description_trip_progress),
                MapboxTripProgressActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_maneuver),
                getString(R.string.description_maneuver),
                MapboxManeuverActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_voice),
                getString(R.string.description_voice),
                MapboxVoiceActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_custom_style),
                getString(R.string.description_custom_style),
                MapboxCustomStyleActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_replay),
                getString(R.string.description_replay),
                ReplayHistoryActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_multiple_arrows),
                getString(R.string.description_multiple_arrows),
                MapboxMultipleArrowActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_independent_route),
                getString(R.string.description_independent_route),
                IndependentRouteGenerationActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_building_highlight),
                getString(R.string.description_building_highlight),
                MapboxBuildingHighlightActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_multileg_route),
                getString(R.string.description_multileg_route),
                MultiLegRouteExampleActivity::class.java
            ),
            SampleItem(
                getString(R.string.title_draw_utility),
                getString(R.string.description_draw_utility),
                RouteDrawingActivity::class.java
            ),
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        locationPermissionsHelper.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onExplanationNeeded(permissionsToExplain: List<String?>?) {
        Toast.makeText(
            this,
            "This app needs location permission in order to show its functionality.",
            Toast.LENGTH_LONG
        ).show()
    }

    override fun onPermissionResult(granted: Boolean) {
        if (granted) {
            requestOptionalPermissions()
        } else {
            Toast.makeText(
                this,
                "You didn't grant location permissions.",
                Toast.LENGTH_LONG
            ).show()
        }
    }

    private fun requestOptionalPermissions() {
        // starting from Android R leak canary writes to Download storage without the permission
        val permissionsToRequest = mutableListOf<String>()
        if (
            Build.VERSION.SDK_INT < Build.VERSION_CODES.R &&
            ContextCompat.checkSelfPermission(this, permission.WRITE_EXTERNAL_STORAGE) !=
            PackageManager.PERMISSION_GRANTED
        ) {
            permissionsToRequest.add(permission.WRITE_EXTERNAL_STORAGE)
        }
        /*if (
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
            ContextCompat.checkSelfPermission(this, permission.POST_NOTIFICATIONS) !=
            PackageManager.PERMISSION_GRANTED
        ) {
            permissionsToRequest.add(permission.POST_NOTIFICATIONS)
        }*/
        if (permissionsToRequest.isNotEmpty()) {
            ActivityCompat.requestPermissions(
                this,
                permissionsToRequest.toTypedArray(),
                10
            )
        }
    }
}
